sklearn/cluster/_spectral.py

```diff
@@ -615,6 +616,34 @@ class SpectralClustering(ClusterMixin, BaseEstimator):
         random_state=0)
     """
 
+    _parameter_constraints = {
+        "n_clusters": [Interval(numbers.Integral, 1, None, closed="left")],
+        "eigen_solver": [StrOptions({"arpack", "lobpcg", "amg"}), None],
+        "n_components": [Interval(numbers.Integral, 1, None, closed="left"), None],
+        "random_state": ["random_state"],
+        "n_init": [Interval(numbers.Integral, 1, None, closed="left")],
+        "gamma": [Interval(numbers.Real, 1.0, None, closed="left")],
+        "affinity": [
+            callable,
+            StrOptions(
+                set(KERNEL_PARAMS.keys())
+                | {"nearest_neighbors", "precomputed", "precomputed_nearest_neighbors"}
+            ),
+        ],
+        "n_neighbors": [Interval(numbers.Integral, 1, None, closed="left")],
+        "eigen_tol": [
+            Interval(numbers.Real, 0.0, None, closed="left"),
+            StrOptions({"auto"}),
+            None,
```

Comment:
`eigen_tol` does not accept `None`

Reply:
Where is that pulled from? I know we pass `eigen_tol=None` for deferred resolution in some places, but I didn't think we did here. Also running the following fails on main (even with the check skipped for `eigen_tol=None`)

```python
from sklearn.cluster import SpectralClustering
import numpy as np
X = np.array([[1, 1], [2, 1], [1, 0],
              [4, 7], [3, 5], [3, 6]])
clustering = SpectralClustering(n_clusters=2,
        assign_labels='discretize',
        random_state=0, eigen_tol=None).fit(X)
```

`eigen_tol` is passed directly from `SpectralClustering-->spectral_clustering-->spectral_embedding-->eigsh` with no resolution of `eigen_tol=None` in that path. The terminal call `eigsh` does not accept `eigen_tol=None`.

Reply:
Ah that is to say the value of `eigen_tol` is set _after_ it is passed by the user. So if the user sets `eigen_tol="auto"` in `SpectralClustering` then **internally** `spectral_embedding` _can_ set `tol=None` to pass to its solvers. In this case the validation we're doing is not on the final value of `tol` but rather what we allow the user to give as argument to `spectral_clustering`. Specifically, we don't want the user to be able to set `eigen_tol=None` because some of the solvers (e.g. `eigsh`) don't accept `None`.